### ISSUE TITLE
.tiff image format is not compatible with the interface

### ISSUE THREAD

**Issue by hanzi4389604** [ai-cfia/nachet-frontend#89]
when uploading a pre-generated .tiff image into the interface, the .tiff image cannot show. Interestingly, the user can still go proceed with the "classify" function and it will generate segmentation and classification results. 

If transform the .tiff into .PNG file, everything works good. 

A .tiff image is now allowed to be attached. Let me know if you need sample .tiff image. Send me the request on teams : liang.zhao@inspection.gc.ca

**Comment by MaxenceGui** [comment]
Issue reported again here :
https://github.com/ai-cfia/nachet-frontend/issues/109#issue-2203424190

Closing the related issue as duplicated
* #109

**Comment by ChromaticPanic** [comment]
Looks like [Safari is the only browser that natively supports Tiff](https://developer.mozilla.org/en-US/docs/Web/Media/Formats/Image_types). So I'm curious if this functionality works if the application is opened in Safari.

I found a few Tiff javascript handling libraries. I can compile some options. Are there any guidelines on choosing? Does this become an ADR?

[Tiff js](https://github.com/seikichi/tiff.js/) js Tiff decoder based on C LibTIFF

[Tiff](https://github.com/image-js/tiff) pure JavaScript image decoder

[Geo Tiff](https://github.com/geotiffjs/geotiff.js/) Pure JS image decoder focusing on Tiff Maps

**Comment by hanzi4389604** [comment]
[087.txt](https://github.com/ai-cfia/nachet-frontend/files/14763970/087.txt)

**Comment by ChromaticPanic** [comment]
I can confirm that the Tiff decoder based on C LibTIFF is able to display Tiff files
![image](https://github.com/ai-cfia/nachet-frontend/assets/78883122/50bb80b7-475c-46f2-83a0-07c42d9fe6c3)

**Comment by ChromaticPanic** [comment]
In terms of implementation there appears to be two options:

Option A: (Likely Less Work and cleaner)
Since PNG is also a lossless format, convert to PNG just for rendering. All the inference will still be on the Tiff format, the original Tiff file remains intact and unaltered during rendering and all saved files will still be Tiff.

Option B: (More Work)
The LoadToCanvas method needs to be rewritten so that if the image is Tiff, we draw the canvas manually.
All of the libraries use the same method to render Tiff directly to the browser

The reason Tiff is not currently displaying is that Image() does not handle it. So image.onload is not triggered when the image is Tiff. So this function would have to be reorganized to process Tiff a different way.

```TypeScript
const loadToCanvas = useCallback((): void => {
    // loads the current image to the canvas and draws the bounding boxes and labels,
    // should update whenever a change is made to the image cache or the score threshold and the selected label is changed
    const image = new Image();
    image.src = imageSrc;
    const canvas: HTMLCanvasElement | null = canvasRef.current;
    if (canvas === null) {
      return;
    }
    const ctx: CanvasRenderingContext2D | null = canvas.getContext("2d");
    if (ctx === null) {
      return;
    }
    image.onload = () => {
      canvas.width = image.width;
      canvas.height = image.height;
      ctx.drawImage(image, 0, 0);
      imageCache.forEach((storedImage) => {
        // find the current image in the image cache based on current index
        if (storedImage.index === imageIndex && storedImage.annotated) {
          storedImage.classifications.forEach((prediction, index) => {
            // !storedImage.overlapping[index]     REMOVE THIS TO SHOW ONLY 1 BB
            if (
              storedImage.scores[index] >= scoreThreshold / 100 &&
              (prediction === selectedLabel || selectedLabel === "all")
            ) {
              ctx.beginPath();
              // draw label index
              ctx.font = "bold 0.9vw Arial";
              ctx.fillStyle = "black";
              ctx.textAlign = "center";
              Object.keys(labelOccurrences).forEach((key, labelIndex) => {
                const scorePercentage = (
                  storedImage.scores[index] * 100
                ).toFixed(0);
                // check to see if label is cut off by the canvas edge, if so, move it to the bottom of the bounding box
                if (storedImage.boxes[index].topY <= 40) {
                  if (prediction === key) {
                    if (switchTable) {
                      ctx.fillText(
                        `[${labelIndex + 1}] - ${scorePercentage}%`,
                        ((storedImage.boxes[index].bottomX as number) -
                          (storedImage.boxes[index].topX as number)) /
                          2 +
                          (storedImage.boxes[index].topX as number),
                        (storedImage.boxes[index].bottomY as number) + 23,
                      );
                    } else {
                      ctx.fillText(
                        `[${index + 1}]`,
                        ((storedImage.boxes[index].bottomX as number) -
                          (storedImage.boxes[index].topX as number)) /
                          2 +
                          (storedImage.boxes[index].topX as number),
                        (storedImage.boxes[index].bottomY as number) + 23,
                      );
                    }
                  }
                } else {
                  // draw label index and score percentage
                  if (prediction === key) {
                    if (switchTable) {
                      ctx.fillText(
                        `[${labelIndex + 1}] - ${scorePercentage}%`,
                        ((storedImage.boxes[index].bottomX as number) -
                          (storedImage.boxes[index].topX as number)) /
                          2 +
                          (storedImage.boxes[index].topX as number),
                        storedImage.boxes[index].topY - 8,
                      );
                    } else {
                      // only draw table if switchTable is false (result component switch button)
                      ctx.fillText(
                        `[${index + 1}]`,
                        ((storedImage.boxes[index].bottomX as number) -
                          (storedImage.boxes[index].topX as number)) /
                          2 +
                          (storedImage.boxes[index].topX as number),
                        storedImage.boxes[index].topY - 8,
                      );
                    }
                  }
                }
              });
              // draw bounding box
              ctx.lineWidth = 2;
              ctx.strokeStyle = "red";
              ctx.rect(
                storedImage.boxes[index].topX,
                storedImage.boxes[index].topY,
                storedImage.boxes[index].bottomX -
                  storedImage.boxes[index].topX,
                storedImage.boxes[index].bottomY -
                  storedImage.boxes[index].topY,
              );
              ctx.stroke();
              ctx.closePath();
            }
          });
        }
        // capture label in bottom left
        if (storedImage.index === imageIndex) {
          storedImage.imageDims = [image.width, image.height];
          ctx.beginPath();
          ctx.font = "bold 0.9vw Arial";
          ctx.textAlign = "left";
          ctx.fillStyle = "#4ee44e";
          ctx.fillText(`Capture ${storedImage.index}`, 10, canvas.height - 15);
          ctx.stroke();
          ctx.closePath();
        }
      });
    };
  }, [
    imageCache,
    imageIndex,
    imageSrc,
    labelOccurrences,
    scoreThreshold,
    selectedLabel,
    switchTable,
  ]);
  ```

From our meeting , I am currently working on Option B

**Comment by rngadam** [comment]
@ChromaticPanic  surprised that since investigation determined Option A is less work and cleaner that we would proceed with Option B instead? Option A sounds good to me. It will probably mean keeping a mapping in the frontend between input tiff and their png for display.

Could you provide sequence diagrams to explain the two options?

https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/

**Comment by ChromaticPanic** [comment]
In terms of implementation there appears to be two options:

Option A: (Likely Less Work and cleaner)
Since PNG is also a lossless format, convert to PNG just for rendering. All the inference will still be on the Tiff format, the original Tiff file remains intact and unaltered during rendering and all saved files will still be Tiff.

Option B: (More Work)
The LoadToCanvas method needs to be rewritten so that if the image is Tiff, we draw the canvas manually.
All of the libraries use the same method to render Tiff directly to the browser

The reason Tiff is not currently displaying is that Image() does not handle it. So image.onload is not triggered when the image is Tiff. So this function would have to be reorganized to process Tiff a different way.

```TypeScript
const loadToCanvas = useCallback((): void => {
    // loads the current image to the canvas and draws the bounding boxes and labels,
    // should update whenever a change is made to the image cache or the score threshold and the selected label is changed
    const image = new Image();
    image.src = imageSrc;
    const canvas: HTMLCanvasElement | null = canvasRef.current;
    if (canvas === null) {
      return;
    }
    const ctx: CanvasRenderingContext2D | null = canvas.getContext("2d");
    if (ctx === null) {
      return;
    }
    image.onload = () => {
      canvas.width = image.width;
      canvas.height = image.height;
      ctx.drawImage(image, 0, 0);
      imageCache.forEach((storedImage) => {
        // find the current image in the image cache based on current index
        if (storedImage.index === imageIndex && storedImage.annotated) {
          storedImage.classifications.forEach((prediction, index) => {
            // !storedImage.overlapping[index]     REMOVE THIS TO SHOW ONLY 1 BB
            if (
              storedImage.scores[index] >= scoreThreshold / 100 &&
              (prediction === selectedLabel || selectedLabel === "all")
            ) {
              ctx.beginPath();
              // draw label index
              ctx.font = "bold 0.9vw Arial";
              ctx.fillStyle = "black";
              ctx.textAlign = "center";
              Object.keys(labelOccurrences).forEach((key, labelIndex) => {
                const scorePercentage = (
                  storedImage.scores[index] * 100
                ).toFixed(0);
                // check to see if label is cut off by the canvas edge, if so, move it to the bottom of the bounding box
                if (storedImage.boxes[index].topY <= 40) {
                  if (prediction === key) {
                    if (switchTable) {
                      ctx.fillText(
                        `[${labelIndex + 1}] - ${scorePercentage}%`,
                        ((storedImage.boxes[index].bottomX as number) -
                          (storedImage.boxes[index].topX as number)) /
                          2 +
                          (storedImage.boxes[index].topX as number),
                        (storedImage.boxes[index].bottomY as number) + 23,
                      );
                    } else {
                      ctx.fillText(
                        `[${index + 1}]`,
                        ((storedImage.boxes[index].bottomX as number) -
                          (storedImage.boxes[index].topX as number)) /
                          2 +
                          (storedImage.boxes[index].topX as number),
                        (storedImage.boxes[index].bottomY as number) + 23,
                      );
                    }
                  }
                } else {
                  // draw label index and score percentage
                  if (prediction === key) {
                    if (switchTable) {
                      ctx.fillText(
                        `[${labelIndex + 1}] - ${scorePercentage}%`,
                        ((storedImage.boxes[index].bottomX as number) -
                          (storedImage.boxes[index].topX as number)) /
                          2 +
                          (storedImage.boxes[index].topX as number),
                        storedImage.boxes[index].topY - 8,
                      );
                    } else {
                      // only draw table if switchTable is false (result component switch button)
                      ctx.fillText(
                        `[${index + 1}]`,
                        ((storedImage.boxes[index].bottomX as number) -
                          (storedImage.boxes[index].topX as number)) /
                          2 +
                          (storedImage.boxes[index].topX as number),
                        storedImage.boxes[index].topY - 8,
                      );
                    }
                  }
                }
              });
              // draw bounding box
              ctx.lineWidth = 2;
              ctx.strokeStyle = "red";
              ctx.rect(
                storedImage.boxes[index].topX,
                storedImage.boxes[index].topY,
                storedImage.boxes[index].bottomX -
                  storedImage.boxes[index].topX,
                storedImage.boxes[index].bottomY -
                  storedImage.boxes[index].topY,
              );
              ctx.stroke();
              ctx.closePath();
            }
          });
        }
        // capture label in bottom left
        if (storedImage.index === imageIndex) {
          storedImage.imageDims = [image.width, image.height];
          ctx.beginPath();
          ctx.font = "bold 0.9vw Arial";
          ctx.textAlign = "left";
          ctx.fillStyle = "#4ee44e";
          ctx.fillText(`Capture ${storedImage.index}`, 10, canvas.height - 15);
          ctx.stroke();
          ctx.closePath();
        }
      });
    };
  }, [
    imageCache,
    imageIndex,
    imageSrc,
    labelOccurrences,
    scoreThreshold,
    selectedLabel,
    switchTable,
  ]);
  ```

From our meeting , I am currently working on Option B

**Comment by ChromaticPanic** [comment]
> @ChromaticPanic surprised that since investigation determined Option A is less work and cleaner that we would proceed with Option B?
> 
> Option A sounds good to me. It will probably mean keeping a mapping in the frontend between input tiff and their png for display.
> 
> Could you provide sequence diagrams to explain the two options?
> 
> https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/

You actually bring up an important downside
Option A would be less refactoring to render png but the cost would be either processing at each re-render (useMemo can do some caching) or we store the pngs in imageCache. Which would double the runtime storage requirements.

Option B would require refactoring that code section but has no additional rendering cost

**Comment by ChromaticPanic** [comment]
Option A.1 with single image useMemo Caching
```mermaid
graph TD;
    loadToCanvas((loadToCanvas))
    splitsInto[Splits into Tiff or Non-Tiff]
    Tiff[Tiff]
    useMemo[useMemo single PNG cache]
    NonTiff[Non-Tiff]
    drawToCanvas[Draw image to canvas]
    drawAnnotations[Draw boxes annotations]
    None[_]

    loadToCanvas --> splitsInto
    splitsInto -->|Tiff| Tiff
    splitsInto -->|Non-Tiff| NonTiff
    Tiff --> useMemo
    useMemo --> drawToCanvas
    NonTiff --> None
    None --> drawToCanvas
    drawToCanvas --> drawAnnotations
```

Option A.2 using imageCache to cache all images
```mermaid
graph TD;
    uploadImage((uploadImage))
    convertToPNG
    imageCache((imageCache))
    loadToCanvas((loadToCanvas))
    splitsInto[Splits into Tiff or Non-Tiff]
    Tiff[Tiff]
    useMemo[grab PNG from cache]
    NonTiff[Non-Tiff]
    drawToCanvas[Draw image to canvas]
    drawAnnotations[Draw boxes annotations]
    None[_]

    uploadImage --> convertToPNG
    convertToPNG --> imageCache
    imageCache --> useMemo
    loadToCanvas --> splitsInto
    splitsInto -->|Tiff| Tiff
    splitsInto -->|Non-Tiff| NonTiff
    Tiff --> useMemo
    useMemo --> drawToCanvas
    NonTiff --> None
    None --> drawToCanvas
    drawToCanvas --> drawAnnotations
```

Option B Refactoring
```mermaid
graph TD;
    loadToCanvas((loadToCanvas))
    splitsInto[Splits into Tiff or Non-Tiff]
    Tiff[Tiff]
    useMemo[Use Tiff lib]
    NonTiff[Non-Tiff]
    drawToCanvas[Draw image to canvas]
    drawAnnotations[Draw boxes annotations]
    None[Use Image lib]

    loadToCanvas --> splitsInto
    splitsInto -->|Tiff| Tiff
    splitsInto -->|Non-Tiff| NonTiff
    Tiff --> useMemo
    useMemo --> drawToCanvas
    NonTiff --> None
    None --> drawToCanvas
    drawToCanvas --> drawAnnotations
```

**Comment by ChromaticPanic** [comment]
Looking at the code more, my impression is that Option A.1 and option B.1 will be similar in the amount of line changes and effort required,

changes will be made in this section
![image](https://github.com/ai-cfia/nachet-frontend/assets/78883122/1f8a1de9-387a-4a86-8a9c-49562db866d9)

Option A.1 will run useMemo before line 462 if the file is tiff then inject the png
Option B.1 will need checking for tiff and using a refactored onload check and instead of ctx.drawImage in line 474 we would use a new tiff canvas drawing method

**Comment by ChromaticPanic** [comment]
![image](https://github.com/ai-cfia/nachet-frontend/assets/78883122/2c79be44-ed1f-464f-8ec3-535493a5acdb)

**Comment by TaranMeyer** [comment]
Not sure where you guys are at with this issue but I am still finding some strange behaviour on the front end:
![image](https://github.com/ai-cfia/nachet-frontend/assets/142623921/f3803066-1dea-427f-be5d-63dd2d39c449)

The image displayed is actually the previous image (Capture 6, a .jpg), with the bounding boxes from capture 7, which is .tif I uploaded, overlaid.

Also, the "correct" bounding boxes didn't show up until I went to view capture 6 and then went back to capture 7. After I hit "classify" what I initially saw was this:

![image](https://github.com/ai-cfia/nachet-frontend/assets/142623921/54733aa9-0cde-4090-aeee-d87da8e4583d)

which is the results for capture 7 but the image of capture 6 and no visible bounding boxes at all.

The classified image should look like this (but of course .tif format which I can't upload here):
![img_20240322_214832](https://github.com/ai-cfia/nachet-frontend/assets/142623921/4f1548ec-82de-45e9-b66a-bfc05d500d34)
